### ISSUE TITLE
Softmax improvement

### ISSUE THREAD

**Issue by bordesf** [Theano/Theano#5712]
I will write some improvement for the softmax implementation as CCW, here is my plan:

- [ ] Allow to have n-dimensional tensor as input where the softmax operation is applied on the last dimension. (Clean some C code and try to parallelize it eventually) 
- [ ] Correct some issues with vector as input (Currently the vector operation works only with the cpu optimization and it returns a matrix instead of a vector)
- [ ] Add a flag to use a spatial softmax in case of 4d tensor (cf https://github.com/Theano/Theano/issues/4269). I think that I will try to match the new implementation with CUDNN where we can choose between the 'instance' mode that corresponds to apply the softmax on the last dimension or the 'channel' mode that allows to perform the softmax on the first dimension.
- [] ...

**Comment by f0k** [comment]
It would be great if one could directly specify the axes to softmax over (accepting an int, a tuple of int, or the string shortcuts 'instance' and 'channel'), and have an optimization that uses the cuDNN implementation whenever possible. But I haven't checked if this is feasible to implement (there's already a jungle of Ops and optimizers just for the simple 2D case). Note that the categorical crossentropy needs to be extended along with the softmax.

**Comment by JesseLivezey** [comment]
FYI, PR #5393 exists and seems to be pretty close to finished although still needs review

**Comment by nouiz** [comment]
That PR just make the current softmax parallel on the CPU. It don't change
anything about what is here.

I should finish it at some point, I want to investigate a few things and do
more detailled profile before merging it.

**Comment by astooke** [comment]
Bump...any updates?

**Comment by nouiz** [comment]
There was much work around, but I tihnk it didn,t got finished/reviewed due
to the of our work on theano.
